### PR TITLE
Prefer to SIGTERM git instead of killing it outright.

### PR DESCRIPTION
**Pull request by vlovich**
On slaves that support it, send the SIGTERM command first & then SIGKILL
after 30 seconds.  On slaves that don't support it, just try to SIGTERM.

This solves the problem where interrupting a git step (or it timing out)
can leave stale lockfiles around.

Resolves http://trac.buildbot.net/ticket/3269

### CODE REVIEW

**Comment by jaredgrubb** [comment]
Tests and notes?

**Comment by vlovich** [comment]
Sure.  What kind of tests are you looking for?

**Comment by jaredgrubb** [comment]
I mean unit tests. We ask that any new features have, at a minimum, test(s) that fail if your code is removed but pass when present, with reasonable coverage. Sometimes that's not possible, but it's what we ask you to try for.

**Comment by vlovich** [comment]
I meant could you elaborate on how I could write such a check.  I'm not sure how to exercise this codepath in unit tests.

File: master/buildbot/steps/source/git.py [hunk 1]
```diff
@@ -352,11 +352,35 @@ def _dovccmd(self, command, abandonOnFailure=True, collectStdout=False, initialS
...
+
+        # check for the interruptSignal flag
+        sigtermTime = None
+        interruptSignal = None
+
+        # If possible prefer to send a SIGTERM to git before we send a SIGKILL.
+        # If we send a SIGKILL, git is prone to leaving around stale lockfiles.
+        # By priming it with a SIGTERM first we can ensure that it has a chance to shut-down gracefully
+        # before getting terminated
+        if not self.slaveVersionIsOlderThan("shell", "2.16"):
```
Comment (sa2ajj) [inline master/buildbot/steps/source/git.py]: You can mock this function to make it return different values.

File: master/buildbot/steps/source/git.py [hunk 2]
```diff
@@ -352,11 +352,35 @@ def _dovccmd(self, command, abandonOnFailure=True, collectStdout=False, initialS
...
+            sigtermTime = '1'
+        else:
+            # Since sigtermTime is unavailable try to just use SIGTERM by itself instead of
+            # killing.  This should be safe.
+            if self.slaveVersionIsOlderThan("shell", "2.15"):
+                log.msg("NOTE: slave does not allow master to specify interruptSignal. This may leave a stale lockfile around if the command is interrupted/times out\n")
+            else:
+                interruptSignal = 'TERM'
+
         cmd = remotecommand.RemoteShellCommand(self.workdir,
```
Comment (sa2ajj) [inline master/buildbot/steps/source/git.py]: and mock this class to check that it's called with correct arguments.

**Comment by sa2ajj** [comment]
HTH

:+1: otherwise!

**Comment by vlovich** [comment]
Is there some example code you can point me at that you would like me to follow the pattern for?
I'm afraid I'd need a lot more time to figure this out or get too frustrated & give up if there's nothing similar.

If there's no existing pattern, is there someone that could write the unit tests for me?  Sorry if I need a lot of hand holding.  I haven't gotten ramped up on the unit tests & they seem very daunting to write.

**Comment by sa2ajj** [comment]
This is how I tested `GerritStatusTarget` in `eight`: [test_status_gerrit.py](https://github.com/buildbot/buildbot/blob/eight/master/buildbot/test/unit/test_status_gerrit.py)

**Comment by vlovich** [comment]
That's probably a good starting point when writing a new test.  However, I'm looking to extend the git tests which seem to follow a very different pattern of writing said tests.  Is there a way to tell expectCommand "hey - I don't care about the remaining commands"?

**Comment by vlovich** [comment]
Added some unit tests.  This was much harder than expected because sigterm & interrupt signal were not being tested anywhere so tests broke all over the place.  I also tried to implement ExpectIgnoreUntil which would be like Expect but ignore any intervening steps but the actual implementation turned out harder than anticipated.  I can split that out into a separate branch if you want.

File: master/buildbot/test/util/steps.py [hunk 3]
```diff
@@ -29,6 +29,50 @@
...
+            missing_in_d2.extend(['{}.{}'.format(k, m) for m in missing_in_v2])
+            for child_k, left, right in different_in_v:
+                different.append(('{}.{}'.format(k, child_k), left, right))
+            continue
+        if d1[k] != d2[k]:
+            different.append((k, d1[k], d2[k]))
+    for k in set(d2.keys()) - both:
+        missing_in_d1.append(k)
+    for k in set(d1.keys()) - both:
+        missing_in_d2.append(k)
```
Comment (jaredgrubb) [inline master/buildbot/test/util/steps.py]: missing_in_d1.extend(set(d2.keys()) - both)
missing_in_d2.extend(set(d1.keys()) - both)

Comment (vlovich) [inline master/buildbot/test/util/steps.py]: Thanks!

**Comment by vlovich** [comment]
I can't tell.  Is the test failure spurious?  Looks to be a network outage.  Not sure where the "recheck" button is to try again.

**Comment by djmitche** [comment]
"zero length field name in format" occurs when using .format with `{}` placeholders on Python-2.6.  The fix is to use `{0}`, `{1}`, etc.

**Comment by vlovich** [comment]
Thanks Dustin.  Updated with this fixed (hopefully).

**Comment by jaredgrubb** [comment]
Your current push only has changes to tests. Is that intentional? I thought there would be a change to Git here?

**Comment by djmitche** [comment]
I think that's everything, then.  I'll have one more look through and then merge.

**Pull request merged**